File: p2/PartitionStream.cpp
// Jomikael Ruiz
// CPSC 3200
// PartitionStream.cpp
// Date Created: Oct. 6th, 2024
// Last Modified: Oct. 15th, 2024

#include "PartitionStream.h"
#include "MsgStream.h"
using namespace std;

/**
 * pre-condition: partitionKey is a valid integer
 */
int PartitionStream::arrayIndexHash(int partitionKey)
{
	return partitionKey % partitionSize;
}

/**
 * pre-condition: partitionKey is a valid integer
 */
void PartitionStream::isValidKey(int partitionKey)
{
	int arrayIndex = arrayIndexHash(partitionKey);



	if( partitionStream[arrayIndex] &&
		partitionKeyStream[arrayIndex] != partitionKey )
	{
		throw(invalid_argument("An Index is occupying the expected partition key"));
	}
}

/**
 * pre-condition: messageStream is a valid MsgStream pointer,
 * 				  partitionKey is a valid integer
 * post-condition: messageStream is inserted into the internal array
 */
void PartitionStream::setMessageStream(int partitionKey, MsgStream * messageStream)
{
	if(!messageStream)
		throw(invalid_argument("MsgStream pointer is nullptr"));

	isValidKey(partitionKey);

	int arrayIndex = arrayIndexHash(partitionKey);

	this->partitionStream[arrayIndex] = messageStream;
	this->partitionKeyStream[arrayIndex] = partitionKey;
	this->isNotEmpty = true;
}

/**
 * pre-condition: src is a valid PartitionStream to deep copy from
 * post-condition: a deep copied version of src is created
 */
void PartitionStream::copyPartition(const PartitionStream& src)
{
	this->partitionSize = src.partitionSize;
	this->isNotEmpty = src.isNotEmpty;
	this->partitionStream = new MsgStream * [partitionSize];
	this->partitionKeyStream = new int[partitionSize];

	for( int i = 0; i < partitionSize; i++ )
	{
		if( !src.partitionStream[i] )
			continue;
		this->partitionStream[i] = new MsgStream( *src.partitionStream[i] );
		this->partitionKeyStream[i] = src.partitionKeyStream[i];
	}
}

/**
 * post-condition: PartitionStream is completely reset to its original form
 */
void PartitionStream::deepReset()
{
	for( unsigned int i = 0; i < partitionSize; i++ )
	{
		delete partitionStream[i];
		partitionStream[i] = nullptr;
	}

	delete [] partitionStream;
	delete [] partitionKeyStream;

	partitionStream = nullptr;
	partitionKeyStream = nullptr;
	this->isNotEmpty = false;
}

PartitionStream::PartitionStream()
{
	partitionSize = DEFAULT_MSGSTREAM_OBJECTS;
	partitionStream = new MsgStream * [partitionSize];
	partitionKeyStream = new int[partitionSize];
	isNotEmpty = false;
}

/**
 * pre-condition: messageStream is a valid MsgStream * for dependency injection
 * 				  partitionKey is a valid integer
 */
PartitionStream::PartitionStream(int partitionKey, MsgStream * messageStream)
{
	partitionSize = DEFAULT_MSGSTREAM_OBJECTS;
	partitionStream = new MsgStream * [partitionSize];
	partitionKeyStream = new int[partitionSize];
	setMessageStream(partitionKey, messageStream);
}

PartitionStream::PartitionStream(const PartitionStream& src)
{
	copyPartition(src);
}

/**
 * pre-condition: src is a valid PartitionStream
 */
PartitionStream::PartitionStream(PartitionStream&& src)
{
	this->partitionSize = src.partitionSize;
	this->isNotEmpty = src.isNotEmpty;
	this->partitionStream = src.partitionStream;
	this->partitionKeyStream = src.partitionKeyStream;

	src.partitionStream = nullptr;
	src.partitionSize = 0;
}

/**
 * post-condition: PartitionStream is completely deallocated
 */
PartitionStream::~PartitionStream()
{
	deepReset();
}

/**
 * pre-condition: src is a valid PartitionStream object
 * post-condition: PartitionStream object is deep copied from src
 */
PartitionStream& PartitionStream::operator=(const PartitionStream& src)
{
	if( this == &src )
		return *this;

	deepReset();
	copyPartition(src);

	return *this;
}

/**
 * pre-condition: src is a valid PartitionStream object
 * post-condition: PartitionStream object is deep copied from src
 */
PartitionStream& PartitionStream::operator=(PartitionStream&& src)
{
	if( this == &src )
		return *this;

	swap(this->partitionSize, src.partitionSize);
	swap(this->isNotEmpty, src.isNotEmpty);
	swap(this->partitionStream, src.partitionStream);
	swap(this->partitionKeyStream, src.partitionKeyStream);

	return *this;
}

/**
 * pre-condition: partitionKey is a valid integer, message is a valid string
 * post-condition: string is added to the respective partition
 */
void PartitionStream::appendMessage(int partitionKey, string message)
{
	int arrayIndex = arrayIndexHash(partitionKey);

	if( !this->partitionStream[arrayIndex] )
	{
		partitionStream[arrayIndex] = new MsgStream();
		partitionKeyStream[arrayIndex] = partitionKey;
	}

	isValidKey(partitionKey);

	partitionStream[arrayIndex]->appendMessage(message);
	isNotEmpty = true;
}

/**
 * pre-condition: partitionKey is a valid integer, messages is a valid MsgStream ptr
 * post-condition: string is added to the respective partition
 */
void PartitionStream::appendMessage(int partitionKey, MsgStream * messages)
{
	if( !messages )
		throw( invalid_argument("MsgStream pointer is nullptr") );
	setMessageStream(partitionKey, messages);
}

/**
 * pre-condition: PartitionStream must not be empty to reset,
 * 				  otherwise an exception is thrown
 * post-condition: internal array is completely wiped and reset
 */
void PartitionStream::resetStream()
{
	if( !isNotEmpty )
		throw(invalid_argument("The Stream is already empty"));

	deepReset();
	this->partitionStream = new MsgStream*[partitionSize];
}

/**
 * pre-condition: the PartitionStream must not be empty,
 * 				  otherwise an exception is thrown,
 * 				  partitionKey must be a valid integer
 * post-condition: a duplicate array with all values in the respective partition
 * 				   is returned in the form of a structure,
 * 				   client is responsible for deallocation of string returned
 */
MsgStream::ArrayReturn PartitionStream::readMessages(int partitionKey)
{
	isValidKey(partitionKey);
	int arrayIndex = arrayIndexHash(partitionKey);

	if( !partitionStream[arrayIndex] )
		throw((invalid_argument("There is no partition allocated here")));

	return partitionStream[arrayIndex]->readAllMessage();
}

File: p3/P3/PartitionStream.cs
namespace P3;

// Jomikael Ruiz
// CPSC 3200
// PartitionStream.cs
// Date Created: 10/21/24
// Last Modified: 10/27/24

/**
 * Class Invariants:
 * - DefaultSize will always be 5 MessageStreams
 * - Max PartitionStream size is 11 MessageStreams
 * - PartitionStream size must be in range 0-11
 * - Client may place DurableStreams in a PartitionStream
 * - Partition Keys will be positive integers with a set of operations performed
 *   to map a partitionKey to its respective index
 * - Partition Keys cannot be negative, otherwise an exception is thrown
 * Immutables:
 * - Once set, PartitionStream size cannot be altered
 */

public class PartitionStream
{
    private const int MaxMsgstreamObjects = 11;
    private const int DefaultMsgstreamObjects = 5;
    private const int DefaultKeyValue = -1;

    private MsgStream[] _partitionStream;
    private int[] _partitionKeyStream;

    private int _partitionSize;
    private bool _isEmpty;

    /**
     * PreCondition: partitionKey is a valid positive integer or 0
     */
    private int ArrayIndexHash(int partitionKey)
    {
        return partitionKey % _partitionSize;
    }

    /**
     * PreCondition: partitionKey is a valid positive integer or 0
     */
    private void IsValidKey(int partitionKey)
    {
        if (partitionKey < 0)
        {
            throw new IndexOutOfRangeException("PartitionKey is a negative value");
        }
        
        int arrayIndex = ArrayIndexHash(partitionKey);
        
        if( _partitionKeyStream[arrayIndex] != -1 &&
            _partitionKeyStream[arrayIndex] != partitionKey )
        {
            throw new IndexOutOfRangeException("An index is already occupying the expected partition");
        }
    }

    /**
     * PreCondition: messageStream is a valid MsgStream or DurableStream,
     * 				  partitionKey is a valid integer
     * PostCondition: messageStream is inserted into the internal array
     */
    private void SetMessageStream(int partitionKey, MsgStream messageStream)
    {
        IsValidKey(partitionKey);

        int arrayIndex = ArrayIndexHash(partitionKey);

        _partitionStream[arrayIndex] = messageStream;
        _partitionKeyStream[arrayIndex] = partitionKey; 
        _isEmpty = false;
    }

    /**
     * PreCondition: PartitionStream must not be empty, otherwise an exception
     *               is thrown
     * PostCondition: PartitionStream is completely reset to its original state by calling
     *                reset() for internal MsgStreams & DurableStreams
     */
    private void Reset()
    {
        if (_isEmpty)
        {
            throw new IndexOutOfRangeException("The stream is empty.");
        }

        for (int i = 0; i < _partitionSize; i++)
        {
            if (_partitionKeyStream[i] == DefaultKeyValue)
                continue;
            _partitionStream[i].Reset();
            _partitionKeyStream[i] = DefaultKeyValue;
        }
        _isEmpty = true;
    }

    /**
     * PreCondition: initialSize is a valid positive integer
     * PostCondition: all internal data for PartitionStream is initialized
     */
    private void InitializePartitionStream(int initialSize)
    {
        _partitionSize = initialSize;
        _partitionStream = new MsgStream[_partitionSize];
        _partitionKeyStream = new int[_partitionSize];
        for (int i = 0; i < _partitionSize; i++)
        {
            _partitionKeyStream[i] = DefaultKeyValue;
        }
        _isEmpty = true;
    }
    
    /**
     * PreCondition: src must be a valid PartitionStream
     * PostCondition: all data from src is DeepCopied into this PartitionStream
     */
    public void DeepCopy(PartitionStream src) {
        _partitionSize = src._partitionSize;
        _isEmpty = src._isEmpty;
        _partitionStream = new MsgStream[_partitionSize];
        _partitionKeyStream = new int[_partitionSize];

        for( int i = 0; i < _partitionSize; i++ )
        {
            if (src._partitionKeyStream[i] == -1)
                continue;
            _partitionStream[i] = new MsgStream( src._partitionStream[i] );
            _partitionKeyStream[i] = src._partitionKeyStream[i];
        }
    }

    public PartitionStream()
    {
        InitializePartitionStream(DefaultMsgstreamObjects);
    }
    
    /**
     * PreCondition: src must be a valid PartitionStream
     */
    public PartitionStream(PartitionStream src)
    {
        DeepCopy(src);
    }
    
    /**
     * PreCondition: inputSize must be a valid positive integer
     */
    public PartitionStream(int inputSize)
    {
        if (inputSize < 0)
            throw new IndexOutOfRangeException("The input size is a negative value");

        if (inputSize > MaxMsgstreamObjects)
            throw new IndexOutOfRangeException("The input size is too large.");
        
        InitializePartitionStream(DefaultMsgstreamObjects);
    }

    /**
     * PreCondition: partitionKey is a valid positive integer, otherwise
     *               an exception is thrown. messageStream must be a valid
     *               MessageStream or DurableStream
     */
    public PartitionStream(int partitionKey, MsgStream messageStream)
    {
        InitializePartitionStream(DefaultMsgstreamObjects);
        SetMessageStream(partitionKey, messageStream);
    }

    /**
     * PreCondition: message is a valid string, partitionKey is a valid integer,
     *               otherwise an exception is thrown
     * PostCondition: messageStream is inserted into the internal array
     */
    public void AppendMessage(int partitionKey, string message)
    {
        int arrayIndex = ArrayIndexHash(partitionKey);

        IsValidKey(partitionKey);

        if (_partitionKeyStream[arrayIndex] == DefaultKeyValue)
        {
            _partitionStream[arrayIndex] = new MsgStream();
            _partitionKeyStream[arrayIndex] = partitionKey;
        }
        _partitionStream[arrayIndex].AppendMessage(message);
        _isEmpty = false;
    }

    /**
     * PreCondition: messageStream is a valid MsgStream or DurableStream,
     * 				  partitionKey is a valid integer
     * PostCondition: messageStream is inserted into the internal array
     */
    public void AppendMessage(int partitionKey, MsgStream messageStream)
    {
        SetMessageStream(partitionKey, messageStream);
    }

    /**
     * PreCondition: PartitionStream must not be empty, otherwise an exception
     *               is thrown
     * PostCondition: PartitionStream is completely reset to its original state
     */
    public void ResetStream()
    {
        Reset();
    }

    /**
     * PreCondition: partitionKey is a valid positive integer,
     *               otherwise an exception is thrown. PartitionStream
     *               must not be empty, otherwise an exception is thrown
     * PostCondition: a string array with the contents of the specific partition
     *                are returned
     */
    public string[] ReadMessages(int partitionKey)
    {
        if (_isEmpty)
            throw new IndexOutOfRangeException("The stream is empty.");
        IsValidKey(partitionKey);
        int arrayIndex = ArrayIndexHash(partitionKey);

        return _partitionStream[arrayIndex].ReadAllMessages();
    }

};

/*
 * Implementation Invariants:
 * - When collisions occur internally with the
 *   partitionKey and arrayIndex, an exception is thrown
 * - Reading messages based on a partitionKey returns all
 *   the messages associated with that specific partition
 * - Object supports Deep Copy via deepCopy() method
 * - ShallowCopy is not supported within this class
 * - To return messages, a structure is returned containing the pointer to
 *   the array of messages and the size for easy traversal
 * - The Key Array is instantly set to values of -1.
 * - PartitionStream supports storage of both MsgStream and DurableStream
 */
